libraries/pbrlib/genglsl/mx_sheen_bsdf.glsl

```diff
@@ -1,89 +1,65 @@
 
-    float NdotV = clamp(dot(N, V), M_FLOAT_EPS, 1.0);
+        vec3 fr = color * mx_imageworks_sheen_brdf(NdotL, NdotV, NdotH, roughness);
+        float dirAlbedo = mx_imageworks_sheen_dir_albedo(NdotV, roughness);
+        bsdf.throughput = vec3(1.0 - dirAlbedo * weight);
 
-    float dirAlbedo = mx_imageworks_sheen_dir_albedo(NdotV, roughness);
-    bsdf.throughput = vec3(1.0 - dirAlbedo * weight);
+        // We need to include NdotL from the light integral here
+        // as in this case it's not cancelled out by the BRDF denominator.
+        bsdf.response = fr * NdotL * occlusion * weight;
+    }
+    else
+    {
+        roughness = clamp(roughness, 0.01, 1.0); // Clamp to range of original impl.
 
-    vec3 Li = mx_environment_irradiance(N);
-    bsdf.response = Li * color * dirAlbedo * weight;
+        vec3 fr = color * mx_zeltner_sheen_brdf(L, V, N, NdotV, roughness);
+        float dirAlbedo = mx_zeltner_sheen_dir_albedo(NdotV, roughness);
+        bsdf.throughput = vec3(1.0 - dirAlbedo * weight);
+        bsdf.response = dirAlbedo * fr * occlusion * weight;
+    }
 }
 
-#elif SHEEN_METHOD == 1
-
-void mx_sheen_bsdf_reflection(vec3 L, vec3 V, vec3 P, float occlusion, float weight, vec3 color, float roughness, vec3 N, inout BSDF bsdf)
+void mx_sheen_bsdf_indirect(vec3 V, float weight, vec3 color, float roughness, vec3 N, int mode, inout BSDF bsdf)
 {
     if (weight < M_FLOAT_EPS)
     {
         return;
     }
 
     N = mx_forward_facing_normal(N, V);
+    float NdotV = clamp(dot(N, V), M_FLOAT_EPS, 1.0);
 
-    float NdotV = min(dot(N, V), 1.0);
-    roughness = clamp(roughness, 0.01, 1.0); // Clamp to range of original impl.
-
-    vec3 fr = color * mx_zeltner_sheen_brdf(L, V, N, NdotV, roughness);
-    float dirAlbedo = mx_zeltner_sheen_dir_albedo(NdotV, roughness);
-    bsdf.throughput = vec3(1.0 - dirAlbedo * weight);
-
-    bsdf.response = dirAlbedo * fr * occlusion * weight;
-}
+    vec3 Li = mx_environment_irradiance(N);
 
-void mx_sheen_bsdf_indirect(vec3 V, float weight, vec3 color, float roughness, vec3 N, inout BSDF bsdf)
-{
-    if (weight < M_FLOAT_EPS)
+    if (mode == 0)
     {
-        return;
+        float dirAlbedo = mx_imageworks_sheen_dir_albedo(NdotV, roughness);
+        bsdf.throughput = vec3(1.0 - dirAlbedo * weight);
+        bsdf.response = Li * color * dirAlbedo * weight;
     }
+    else
+    {
+        roughness = clamp(roughness, 0.01, 1.0); // Clamp to range of original impl.
 
-    N = mx_forward_facing_normal(N, V);
-
-    float NdotV = min(dot(N, V), 1.0);
-    roughness = clamp(roughness, 0.01, 1.0); // Clamp to range of original impl.
-
-    float dirAlbedo = mx_zeltner_sheen_dir_albedo(NdotV, roughness);
-    bsdf.throughput = vec3(1.0 - dirAlbedo * weight);
-
-    vec3 Li = mx_environment_irradiance(N);
-    bsdf.response = Li * color * dirAlbedo * weight;
+        float dirAlbedo = mx_zeltner_sheen_dir_albedo(NdotV, roughness);
+        bsdf.throughput = vec3(1.0 - dirAlbedo * weight);
```

Comment:
Since these lines are the same as in the if clause, you could move them to the end (and make `dirAlbedo` non-local).